### DISCUSSION TITLE
barcode scan and amazon lookup application

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [zauberzeug/nicegui#2028] [Show and tell]
I have a bunch of Bluerays i don't need any more and want to offer to my friends and family. To compile a list of these i thought of
scanning in the bar codes and looking up the results. There are online tools for this and apps like google lense. I figured it might
be fun to try to integrate this usecase  into my existing [scan2wiki](https://github.com/WolfgangFahl/scan2wiki) nicegui application:

So i just implemented 
* [add webcam scan](https://github.com/WolfgangFahl/scan2wiki/issues/13) using e.g [IP Webcam android smart phone a pp](https://play.google.com/store/apps/details?id=com.pas.webcam&hl=de&gl=US)
* [add barcode detection feature](https://github.com/WolfgangFahl/scan2wiki/issues/14)
* [add barcode feature to webcam form](https://github.com/WolfgangFahl/scan2wiki/issues/15)
* [show list of products scanned so far and save and load to json file](https://github.com/WolfgangFahl/scan2wiki/issues/16)

The screenshot below shows the result:
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/f5b8a108-22b3-4cb6-915d-6b8b8ff48494)

The lookup gives:
```
Code: 5051890305695, Type: EAN13, Product: Game of Thrones - Staffel 6 [Blu-ray], Price: 16,67 €, ASIN: B01JVQJK18
```
which correctly points to Games of Thrones season 6 👍 

https://www.amazon.de/dp/B01JVQJK18

With a bit of html the amazon link can be nicely shown as a thumbnail:
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/3cf90117-a32a-42ca-84d4-1bee57a9b10f)

and with a bit more code a list of products scanned so far can be shown and new scans be added and stored as json:
```json
[
  {
    "title": "The Beatles - A Hard Day's Night",
    "image_url": "https://m.media-amazon.com/images/I/81m01dZR2UL._AC_UY218_.jpg",
    "price": "4,99 \u20ac",
    "asin": "B00KHK1SW2",
    "ean": "4020628887711"
  },

....
```
![grafik](https://github.com/WolfgangFahl/scan2wiki/assets/1336221/e238684f-075f-47b4-86c9-1cae7bdc7bb3)

I thinks this a great example of how quickly you can get useful apps with nicegui. I also used [ngwidgets](https://github.com/WolfgangFahl/nicegui_widgets) to not have to bother with menu, version and other basic things that are needed for all apps

**Comment by alikalik9** [comment]
Great project!

**Comment by WolfgangFahl** [comment]
@alikalik9 thx for the comment i'd appreciate a star at https://github.com/WolfgangFahl/scan2wiki

**Comment by WolfgangFahl** [comment]
# Installation
```bash
pip install scan2wiki
```
# zbar prerequisite
see install script - please give feedback in the discussion and/or issues of the project for non supported environments. Since i do not use Windows i have no clue how to install the zbar library on Windows. May be we need a dockerized version in this case.
```bash
#!/bin/bash
# WF 2023-11-16
# Updated to include conditional installation for zbar

# Upgrade pip
pip install --upgrade pip

# Install Python dependencies from your project
pip install .

# Check the operating system
OS_NAME=$(uname -s)
echo "installing scan2wiki on $OS_NAME"
case "$OS_NAME" in
    Darwin)
        # macOS
        if command -v brew >/dev/null 2>&1; then
            # Install zbar using Homebrew
            brew install zbar
        elif command -v port >/dev/null 2>&1; then
            # Install zbar using MacPorts
            sudo port install zbar
        else
            echo "Error: Homebrew or MacPorts is required to install zbar on macOS."
            exit 1
        fi
        ;;

    Linux)
        # Linux (assuming Debian-based systems)
        sudo apt-get update
        sudo apt-get install -y libzbar0
        ;;

    MINGW*|MSYS*|CYGWIN*)
        # Windows (assuming usage of Git Bash or similar environment)
        # For actual Windows command line, consider using Chocolatey
        echo "Please manually install zbar for Windows or use Chocolatey in a PowerShell or CMD script."
        ;;

    *)
        echo "Unsupported OS: $OS_NAME"
        exit 1
        ;;
esac
```

# Usage
```
usage: scan2wiki [-h] [-a] [-c] [-d] [--debugServer DEBUGSERVER]
                 [--debugPort DEBUGPORT] [--debugRemotePath DEBUGREMOTEPATH]
                 [--debugLocalPath DEBUGLOCALPATH] [-l] [-i INPUT] [-rol]
                 [--host HOST] [--port PORT] [-s] [-V] [-v] [-rp ROOT_PATH]
                 [-wc WEBCAM]

Scan to Wiki by watching a scan folder

options:
  -h, --help            show this help message and exit
  -a, --about           show about info [default: False]
  -c, --client          start client [default: False]
  -d, --debug           show debug info [default: False]
  --debugServer DEBUGSERVER
                        remote debug Server
  --debugPort DEBUGPORT
                        remote debug Port
  --debugRemotePath DEBUGREMOTEPATH
                        remote debug Server path mapping - remotePath - path
                        on debug server
  --debugLocalPath DEBUGLOCALPATH
                        remote debug Server path mapping - localPath - path on
                        machine where python runs
  -l, --local           run with local file system access [default: False]
  -i INPUT, --input INPUT
                        input file
  -rol, --render_on_load
                        render on load [default: False]
  --host HOST           the host to serve / listen from [default: localhost]
  --port PORT           the port to serve from [default: 8334]
  -s, --serve           start webserver [default: False]
  -V, --version         show program's version number and exit
  -v, --verbose         show verbose output [default: False]
  -rp ROOT_PATH, --root_path ROOT_PATH
                        path to example pdf files [default:
                        /Users/wf/Library/Python/3.10/lib/python/site-
                        packages/scan2wiki_examples]
  -wc WEBCAM, --webcam WEBCAM
                        url of webcam for scans [default: None]
```

# Start 
```bash
scan2wiki -s -c
```

set the -wc option to the url of your webcam if you'd like to scan

set the --port for a different port

set --host to 0.0.0.0 if you'd like to access the server from any other host in your intranet

8334 is the default port so
http://localhost:8334/webcam should make you a happy camper for trying things out.

# products.json 
If you'd like to start with a producst.json put in in $HOME/.scan2wiki/products.json e.g
```json
[
  {
    "title": "The Beatles - A Hard Day's Night",
    "image_url": "https://m.media-amazon.com/images/I/81m01dZR2UL._AC_UY218_.jpg",
    "price": "4,99 \u20ac",
    "asin": "B00KHK1SW2",
    "ean": "4020628887711"
  }
]
```

# Issues and PRs 
Of course issues and PRs are welcome!
